autoblocks/_impl/global_state.py

```diff
@@ -1,20 +1,105 @@
 
 import httpx
 
+from autoblocks._impl.util import AnyTask
+
 log = logging.getLogger(__name__)
 
+_started: bool = False
+_thread: Optional[threading.Thread] = None
+_loop: Optional[asyncio.AbstractEventLoop] = None
 _client: Optional[httpx.AsyncClient] = None
 _sync_client: Optional[httpx.Client] = None
-_loop: Optional[asyncio.AbstractEventLoop] = None
-_started: bool = False
+_background_tasks: Set[AnyTask] = set()
+_main_thread_has_finished: bool = False
+
+
+def _run_event_loop(_event_loop: asyncio.AbstractEventLoop) -> None:
+    asyncio.set_event_loop(_event_loop)
+    _event_loop.run_forever()
+
+
+def flush(timeout: Optional[timedelta] = None) -> None:
+    """
+    Wait for all pending tasks to complete.
+    """
+    timeout_seconds = timeout.total_seconds() if timeout else 30
+
+    log.debug("Flushing background tasks with timeout of % seconds", timeout_seconds)
+    if not _background_tasks:
+        # Already empty
+        log.debug("No background tasks to flush")
+        return
+
+    start_time = time.time()
+    log.debug("Waiting for %s background tasks to finish...", len(_background_tasks))
+    while _background_tasks and (time.time() - start_time) < timeout_seconds:
+        time.sleep(0.1)
+
+    if _background_tasks:
+        log.error(
+            "Timed out waiting for background tasks to flush. % tasks left unfinished.",
+            len(_background_tasks),
+        )
+    else:
+        log.debug("Successfully flushed all background tasks")
+
+
+def _flush_and_shut_down_event_loop() -> None:
+    """
+    Flushes all background tasks and then shuts down the event loop.
+    """
+    flush()
+
+    if _thread and _loop and _loop.is_running():
+        # Stop the event loop (will cause run_forever to stop)
+        log.debug("Stopping event loop")
+        _loop.call_soon_threadsafe(_loop.stop)
+        # Wait for the thread to finish (will happen when run_forever stops)
+        log.debug("Waiting for background thread to finish")
+        _thread.join()
+        # Cancel all remaining tasks
+        log.debug("Cancelling all remaining tasks")
+        _loop.run_until_complete(_loop.shutdown_asyncgens())
+        # Close the loop
+        log.debug("Closing event loop")
+        _loop.close()
+        log.debug("Event loop closed")
+
+
+def _main_shut_down_monitor_thread() -> None:
+    """
+    Start a thread that waits for the main thread to finish and
+    then flushes + shuts down the event loop.
+
+    See https://stackoverflow.com/a/63075281
+    """
+    main_thread = threading.main_thread()
+    main_thread.join()
+
+    # Used by the tracer to switch to the sync HTTP client
+    # after the interpreter has shut down. Otherwise we'll get:
+    # RuntimeError: cannot schedule new futures after interpreter shutdown
+    # TODO: figure out how to keep the main thread alive while
+    # we flush the tasks so that we don't need to switch to the
+    # sync client. Users will also run into issues if their
```

Comment:
Just a note to myself: Conceptually - All we should have to do is not use a daemon thread for the background worker - because in that case the main thread won't exit until all other threads are joined.

I went down this path - We both ended up at the same error though (interpreter shutdown)

Reply:
> because in that case the main thread won't exit until all other threads are joined

are you sure? i think we wouldn't need the workaround if this were true, because then we would not see the "cannot schedule new futures" error while our non-daemon background thread is still running

Reply:
These are the docs that made me think this

"
A thread can be flagged as a “daemon thread”. The significance of this flag is that the entire Python program exits when only daemon threads are left. The initial value is inherited from the creating thread. The flag can be set through the [daemon](https://docs.python.org/3/library/threading.html#threading.Thread.daemon) property or the daemon constructor argument.

Note Daemon threads are abruptly stopped at shutdown. Their resources (such as open files, database transactions, etc.) may not be released properly. If you want your threads to stop gracefully, make them non-daemonic and use a suitable signalling mechanism such as an [Event](https://docs.python.org/3/library/threading.html#threading.Event).
"

I think the issue is how `main` spawns the interpreter (as defined in https://docs.python.org/3/library/threading.html#threading.main_thread) that other threads use?

Reply:
i think it's actually a bug that started in python 3.9 wrt ThreadPoolExecutor: https://bugs.python.org/issue42647